### PR TITLE
Redis

### PR DESCRIPTION
**Pull request by ajax-malov-n**
Added
- Redis

### CODE REVIEW

File: domainservice/src/main/kotlin/systems/ajax/malov/stockanalyzer/repository/CacheStockRecordRepository.kt [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package systems.ajax.malov.stockanalyzer.repository
+
+import reactor.core.publisher.Flux
+import reactor.core.publisher.Mono
+import systems.ajax.malov.stockanalyzer.entity.MongoStockRecord
+import java.util.Date
+
+interface CacheStockRecordRepository {
```
Comment (ajax-gnatiuk-s) [inline domainservice/src/main/kotlin/systems/ajax/malov/stockanalyzer/repository/CacheStockRecordRepository.kt]: The interface doesn't have any data related to caching. So, the word `Cache` in the interface name is irrelevant.

You have `StockRecordRepository` that has the same functions + `insertAll` extra function.

Maybe, renaming to `ReadOnlyStockRepordRepository` would be more suitable.
Moreover, you can extend interfaces:
```kotlin
interface StockRecordRepository : ReadOnlyStockRepordRepository {
    // extra function: insertAll
}
```

File: domainservice/src/main/kotlin/systems/ajax/malov/stockanalyzer/config/RedisConfig.kt [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+import org.springframework.data.redis.connection.lettuce.LettuceConnectionFactory
+import org.springframework.data.redis.core.ReactiveRedisTemplate
+import org.springframework.data.redis.serializer.Jackson2JsonRedisSerializer
+import org.springframework.data.redis.serializer.RedisSerializationContext
+import org.springframework.data.redis.serializer.StringRedisSerializer
+import systems.ajax.malov.stockanalyzer.entity.MongoStockRecord
+import java.time.Duration
+
+@Configuration
+@EnableCaching
```
Comment (ajax-gnatiuk-s) [inline domainservice/src/main/kotlin/systems/ajax/malov/stockanalyzer/config/RedisConfig.kt]: why do we need this annotation?

File: domainservice/src/main/kotlin/systems/ajax/malov/stockanalyzer/repository/impl/RedisStockRecordRepository.kt [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+                log.error(it.message, it)
+                stockMongoRecordRepository.findTopStockSymbolsWithStockRecords(quantity, from, to)
+            }.onErrorResume(RedisCommandTimeoutException::class.java) {
+                log.error(it.message, it)
+                stockMongoRecordRepository.findTopStockSymbolsWithStockRecords(quantity, from, to)
+            }
+            .onErrorResume(SocketException::class.java) {
+                log.error(it.message, it)
+                stockMongoRecordRepository.findTopStockSymbolsWithStockRecords(quantity, from, to)
+            }
```
Comment (ajax-gnatiuk-s) [inline domainservice/src/main/kotlin/systems/ajax/malov/stockanalyzer/repository/impl/RedisStockRecordRepository.kt:43]:
```
.onErrorResume(<EXCEPTION>::class.java) {
                log.error(it.message, it)
                stockMongoRecordRepository.findTopStockSymbolsWithStockRecords(quantity, from, to)
            }
```
is duplicated three times. Can we do something with it?

File: domainservice/src/main/kotlin/systems/ajax/malov/stockanalyzer/repository/impl/RedisStockRecordRepository.kt [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+                stockMongoRecordRepository.findAllStockSymbols()
+            }
+            .onErrorResume(RedisCommandTimeoutException::class.java) {
+                log.error(it.message, it)
+                stockMongoRecordRepository.findAllStockSymbols()
+            }
+            .onErrorResume(SocketException::class.java) {
+                log.error(it.message, it)
+                stockMongoRecordRepository.findAllStockSymbols()
+            }
```
Comment (ajax-gnatiuk-s) [inline domainservice/src/main/kotlin/systems/ajax/malov/stockanalyzer/repository/impl/RedisStockRecordRepository.kt:57]: Avoid code duplication

File: domainservice/src/main/kotlin/systems/ajax/malov/stockanalyzer/repository/CacheStockRecordRepository.kt (hunk 1 above)
Comment (ajax-gnatiuk-s) [reply domainservice/src/main/kotlin/systems/ajax/malov/stockanalyzer/repository/CacheStockRecordRepository.kt]: Actually, we can use the same interface for both: Mongo and Redis

File: domainservice/src/main/kotlin/systems/ajax/malov/stockanalyzer/config/RedisConfig.kt (hunk 2 above)
Comment (ajax-malov-n) [reply domainservice/src/main/kotlin/systems/ajax/malov/stockanalyzer/config/RedisConfig.kt]: We don`t need it)

File: domainservice/src/main/kotlin/systems/ajax/malov/stockanalyzer/repository/impl/RedisStockRecordRepository.kt [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            .log()
+            .onErrorResume(::isRedisOrSocketException) {
+                log.error(it.message, it)
+                stockMongoRecordRepository.findTopStockSymbolsWithStockRecords(quantity, from, to)
+            }
+            .switchIfEmpty(
+                stockMongoRecordRepository.findTopStockSymbolsWithStockRecords(quantity, from, to).flatMap {
+                    saveTopStockSymbolsWithStockRecords(quantity, it)
+                }
+            ).log()
```
Comment (ajax-gnatiuk-s) [inline domainservice/src/main/kotlin/systems/ajax/malov/stockanalyzer/repository/impl/RedisStockRecordRepository.kt]: do we still need `log()`?

File: domainservice/src/main/kotlin/systems/ajax/malov/stockanalyzer/repository/impl/RedisStockRecordRepository.kt [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    private val stockMongoRecordRepository: StockRecordRepository,
+) : ReadOnlyStockRecordRepository {
+    override fun findTopStockSymbolsWithStockRecords(
+        quantity: Int,
+        from: Date,
+        to: Date,
+    ): Mono<Map<String, List<MongoStockRecord>>> {
+        return reactiveMapRedisTemplate.opsForValue()
+            .get(createTopStockSymbolsWithStockRecordsKey(quantity, stockRecordPrefix))
+            .log()
```
Comment (ajax-gnatiuk-s) [inline domainservice/src/main/kotlin/systems/ajax/malov/stockanalyzer/repository/impl/RedisStockRecordRepository.kt]: delete this?

**Review by ajax-gnatiuk-s — APPROVED** [review]
lgtm

**Pull request merged**